Panic in local store

The local store panicked today twice in my setup (see prt screen below).

It happened on an empty data directory, maybe 10 seconds after I started the node, just after I shut down my node.

It can't be reproduced reliably.


![image.png](https://images.zenhubusercontent.com/5e7c6d2ce969e644bdebbf30/61e16e33-74ca-4f2e-8b46-7ea068f7603c)

This happened again to me today, now when shutting down a node which has been running for a while (8+ hours)

Logs:

```
TRAC[2020-06-25T10:04:18+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:04:21+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65: read delivery: context deadline exceeded peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
TRAC[2020-06-25T10:04:21+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
DEBU[2020-06-25T10:04:22+02:00] retrieval: failed to get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb: read delivery: context deadline exceeded peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
TRAC[2020-06-25T10:04:22+02:00] retrieval: get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc 
DEBU[2020-06-25T10:04:22+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb: read delivery: EOF peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
TRAC[2020-06-25T10:04:22+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
DEBU[2020-06-25T10:04:28+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
TRAC[2020-06-25T10:04:28+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:04:31+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362: read delivery: context deadline exceeded peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
TRAC[2020-06-25T10:04:31+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:04:32+02:00] retrieval: failed to get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc: read delivery: context deadline exceeded peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc 
TRAC[2020-06-25T10:04:32+02:00] retrieval: get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
DEBU[2020-06-25T10:04:32+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65: read delivery: context deadline exceeded peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
TRAC[2020-06-25T10:04:32+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
DEBU[2020-06-25T10:04:38+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: read delivery: context deadline exceeded peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
TRAC[2020-06-25T10:04:38+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9 
DEBU[2020-06-25T10:04:41+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: read delivery: context deadline exceeded peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
TRAC[2020-06-25T10:04:41+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:04:42+02:00] retrieval: failed to get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65: read delivery: context deadline exceeded peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
TRAC[2020-06-25T10:04:42+02:00] retrieval: get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
DEBU[2020-06-25T10:04:42+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362: read delivery: context deadline exceeded peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
TRAC[2020-06-25T10:04:42+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:04:48+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9: read delivery: context deadline exceeded peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9 
TRAC[2020-06-25T10:04:48+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319 
DEBU[2020-06-25T10:04:51+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:04:51+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:04:51+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
TRAC[2020-06-25T10:04:51+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc 
DEBU[2020-06-25T10:04:52+02:00] retrieval: failed to get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362: read delivery: context deadline exceeded peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
TRAC[2020-06-25T10:04:52+02:00] retrieval: get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:04:52+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:04:52+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:04:52+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:04:58+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319: read delivery: context deadline exceeded peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319 
TRAC[2020-06-25T10:04:58+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 
DEBU[2020-06-25T10:05:01+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc: read delivery: context deadline exceeded peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc 
TRAC[2020-06-25T10:05:01+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
DEBU[2020-06-25T10:05:02+02:00] retrieval: failed to get chunk 2d4c6571e2527a7488b1f6d23fefe2f35200b737a42e89b3dac0b8f5b6aa9a15 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: read delivery: context deadline exceeded peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:05:02+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:05:02+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
TRAC[2020-06-25T10:05:02+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc 
DEBU[2020-06-25T10:05:02+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc: read delivery: EOF peer 15580e787a18c0efd336edcb84db24458c98da4e83d44e88294b9d5fc5669fbc 
TRAC[2020-06-25T10:05:02+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
INFO[2020-06-25T10:05:07+02:00] debug api access                              duration=0.020260202 ip="::1" method=GET proto=HTTP/1.1 size=4037 status=200 uri=/topology user-agent=curl/7.70.0
DEBU[2020-06-25T10:05:08+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer c089ca1e67ab90ee4f1d640b4a51054a448b18d8cf0e641375116ad875b112de: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer c089ca1e67ab90ee4f1d640b4a51054a448b18d8cf0e641375116ad875b112de 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer c089ca1e67ab90ee4f1d640b4a51054a448b18d8cf0e641375116ad875b112de: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer c089ca1e67ab90ee4f1d640b4a51054a448b18d8cf0e641375116ad875b112de 
DEBU[2020-06-25T10:05:08+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9: error: get from store: netstore retrieve chunk: read delivery: context deadline exceeded peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9 
DEBU[2020-06-25T10:05:11+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65: read delivery: context deadline exceeded peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
TRAC[2020-06-25T10:05:11+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
DEBU[2020-06-25T10:05:12+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb: read delivery: EOF peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
TRAC[2020-06-25T10:05:12+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
TRAC[2020-06-25T10:05:15+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
^CDEBU[2020-06-25T10:05:17+02:00] received signal: interrupt                   
INFO[2020-06-25T10:05:17+02:00] shutting down                                
DEBU[2020-06-25T10:05:17+02:00] localstore push subscription iteration: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 6, since: 2102, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 10, since: 141, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 12, since: 35, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362: read delivery: context canceled peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65: read delivery: context canceled peer 4cfee2a66c8e0e6d0f9ccd2519f5af3058af49702fe76fb1ea55ac46382a8a65 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm8CFyLbRxVQaTvh45RESGWPzs65QtCXPnfAx7QaZMznaY": context canceled 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 25e21ed5b3cc57e1a332883b9e14d98d27b0f99887a7b8cd6a712038c990ae61 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAkwqFGJzeConsi2qZqNMXAqKCi7dDoNCNHhHgVVYhMqzwd": context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb: error: get from store: netstore retrieve chunk: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAkwqFGJzeConsi2qZqNMXAqKCi7dDoNCNHhHgVVYhMqzwd": context canceled peer 2518d4353b658b26b9b4b26d36ecae827f28ca02a45451ea789678710b0b6beb 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: error: get from store: netstore retrieve chunk: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAkwqFGJzeConsi2qZqNMXAqKCi7dDoNCNHhHgVVYhMqzwd": context canceled peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:05:17+02:00] kademlia dialing to peer 916bed288d96ea37bf153ca081cb8e67d62921dedf8cca398722665429f13cfd 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 3, since: 13757, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 4, since: 8773, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 7, since: 1109, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 8, since: 530, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 11, since: 73, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 9, since: 282, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 14, since: 10, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 15, since: 13, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 5, since: 4531, until: 18446744073709551615: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: read delivery: context canceled peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm8CFyLbRxVQaTvh45RESGWPzs65QtCXPnfAx7QaZMznaY": context canceled 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer d6a73e493ec42fca92b78ee980e3e9f254527dfe533708102daf31bb48bb9cd9: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm3pRWxBQJoeLqQu2A4wmzAZBf8Q4fu45FQVQ3aApubuwq": context canceled 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer c089ca1e67ab90ee4f1d640b4a51054a448b18d8cf0e641375116ad875b112de 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer c089ca1e67ab90ee4f1d640b4a51054a448b18d8cf0e641375116ad875b112de: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAmB4phowheHdKCF9UvbKqULTb6FUD1iiuzs88RudtMxArT": context canceled 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 969dae6c2596bd59109601b410c32dd2f4637c033f62496e69760239a14407f2 from peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm6Pqd41oVvMpqaHeHxbkby6UEpXn9hv2JGTRYi1Shj4zY": context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717: error: get from store: netstore retrieve chunk: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm6Pqd41oVvMpqaHeHxbkby6UEpXn9hv2JGTRYi1Shj4zY": context canceled peer efd462c156fd1da8ed6071ce190855db657062d25c5fbab7fcd869a672e8e717 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol pullsync/1.0.0: stream pullsync: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: make offer: context canceled 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer 539e24a1837ba0d8c0b39f7b261dab1cc5be10e9b2563b24638f092970752362: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAmJbafufHyqATbcmXSUz1JCJ99Tg39TKF8TjJn6k4BAqAG": context canceled 
TRAC[2020-06-25T10:05:17+02:00] retrieval: get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319 
DEBU[2020-06-25T10:05:17+02:00] retrieval: failed to get chunk 0eb3c1e3b43d91b6c0c1d39965809313407f7c58be02cf7c4ff73fb99d57568a from peer ca562507d71bb61a2e74d8d2445a0f542d7f34512827a914f9d837f6bc11f319: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm6Pqd41oVvMpqaHeHxbkby6UEpXn9hv2JGTRYi1Shj4zY": context canceled 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10: error: get from store: netstore retrieve chunk: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm6Pqd41oVvMpqaHeHxbkby6UEpXn9hv2JGTRYi1Shj4zY": context canceled peer bb46e0540755cea50b409b4f2bb3682d704119686dd81612b122a5cdd15ccd10 
DEBU[2020-06-25T10:05:17+02:00] error handle protocol retrieval/1.0.0: stream retrieval: peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa: error: get from store: netstore retrieve chunk: new stream: create stream "/swarm/retrieval/1.0.0/retrieval" to "16Uiu2HAm6Pqd41oVvMpqaHeHxbkby6UEpXn9hv2JGTRYi1Shj4zY": context canceled peer 9466f95170d11cc3543182e37763d1867a601bea336902118ff1f43081003aaa 
DEBU[2020-06-25T10:05:17+02:00] localstore pull subscription iteration: bin: 13, since: 17, until: 18446744073709551615: context canceled 
panic: runtime error: slice bounds out of range [:16] with capacity 0

goroutine 417343 [running]:
github.com/ethersphere/bee/pkg/localstore.New.func4(0xc001e35380, 0x20, 0x30, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/ethersphere/bee/pkg/localstore/localstore.go:233 +0xd1
github.com/ethersphere/bee/pkg/shed.Index.Get(0xc0001425b0, 0xc000203f79, 0x1, 0x1, 0xc00020e780, 0xc000e11f80, 0x11e59e0, 0x11e59e8, 0xc001e35380, 0x20, ...)
	github.com/ethersphere/bee/pkg/shed/index.go:149 +0x25d
github.com/ethersphere/bee/pkg/localstore.(*DB).setSync(0xc000286600, 0xc001d38100, 0xc001e35380, 0x20, 0x30, 0x2, 0x1b79d20, 0x331b, 0xc000e9db00)
	github.com/ethersphere/bee/pkg/localstore/mode_set.go:214 +0x162
github.com/ethersphere/bee/pkg/localstore.(*DB).set(0xc000286600, 0x2, 0xc001c05200, 0x21, 0x40, 0x0, 0x0)
	github.com/ethersphere/bee/pkg/localstore/mode_set.go:82 +0x1b1
github.com/ethersphere/bee/pkg/localstore.(*DB).Set(0xc000286600, 0x133d7c0, 0xc0002ef840, 0x2, 0xc001c05200, 0x21, 0x40, 0x0, 0x0)
	github.com/ethersphere/bee/pkg/localstore/mode_set.go:38 +0xf3
github.com/ethersphere/bee/pkg/pullsync.(*Syncer).setChunks(...)
	github.com/ethersphere/bee/pkg/pullsync/pullsync.go:242
github.com/ethersphere/bee/pkg/pullsync.(*Syncer).handler(0xc00035c370, 0x133d7c0, 0xc0002ef840, 0xc002b18940, 0x20, 0x20, 0x13405c0, 0xc0003562c0, 0x0, 0x0)
	github.com/ethersphere/bee/pkg/pullsync/pullsync.go:232 +0x89f
github.com/ethersphere/bee/pkg/p2p/libp2p.(*Service).AddProtocol.func1(0x134a9e0, 0xc0003562a0)
	github.com/ethersphere/bee/pkg/p2p/libp2p/libp2p.go:306 +0xb90
github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).SetStreamHandlerMatch.func1(0xc002ac7600, 0x1e, 0x7feca90c3ce0, 0xc0003562a0, 0x1e, 0xc0002a5eb0)
	github.com/libp2p/go-libp2p@v0.5.1/p2p/host/basic/basic_host.go:411 +0x9d
created by github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).newStreamHandler
	github.com/libp2p/go-libp2p@v0.5.1/p2p/host/basic/basic_host.go:295 +0x6a
```

--- a/pkg/puller/puller.go
+++ b/pkg/puller/puller.go
@@ -49,6 +49,7 @@ type Puller struct {
 	cursorsMtx sync.Mutex
 
 	quit chan struct{}
+	wg   sync.WaitGroup
 }
 
 func New(o Options) *Puller {
@@ -62,12 +63,13 @@ func New(o Options) *Puller {
 
 		syncPeers: make([]map[string]*syncPeer, bins),
 		quit:      make(chan struct{}),
+		wg:        sync.WaitGroup{},
 	}
 
 	for i := uint8(0); i < bins; i++ {
 		p.syncPeers[i] = make(map[string]*syncPeer)
 	}
-
+	p.wg.Add(1)
 	go p.manage()
 	return p
 }
@@ -78,11 +80,15 @@ type peer struct {
 }
 
 func (p *Puller) manage() {
+	defer p.wg.Done()
 	c, unsubscribe := p.topology.SubscribePeersChange()
 	defer unsubscribe()
 
 	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	go func() {
+		<-p.quit
+		cancel()
+	}()
 	for {
 		select {
 		case <-c:
@@ -229,8 +235,10 @@ func (p *Puller) recalcPeer(ctx context.Context, peer swarm.Address, po, d uint8
 				binCtx, cancel := context.WithCancel(ctx)
 				syncCtx.binCancelFuncs[bin] = cancel
 				if cur > 0 {
+					p.wg.Add(1)
 					go p.histSyncWorker(binCtx, peer, bin, cur)
 				}
+				p.wg.Add(1)
 				go p.liveSyncWorker(binCtx, peer, bin, cur)
 			}
 		}
@@ -263,8 +271,10 @@ func (p *Puller) recalcPeer(ctx context.Context, peer swarm.Address, po, d uint8
 			binCtx, cancel := context.WithCancel(ctx)
 			syncCtx.binCancelFuncs[po] = cancel
 			if cur > 0 {
+				p.wg.Add(1)
 				go p.histSyncWorker(binCtx, peer, want, cur)
 			}
+			p.wg.Add(1)
 			go p.liveSyncWorker(binCtx, peer, want, cur)
 		}
 		for _, bin := range dontWant {
@@ -310,8 +320,10 @@ func (p *Puller) syncPeer(ctx context.Context, peer swarm.Address, po, d uint8) 
 		binCtx, cancel := context.WithCancel(ctx)
 		syncCtx.binCancelFuncs[po] = cancel
 		if cur > 0 {
+			p.wg.Add(1)
 			go p.histSyncWorker(binCtx, peer, bin, cur) // start historical
 		}
+		p.wg.Add(1)
 		go p.liveSyncWorker(binCtx, peer, bin, cur) // start live
 
 		return
@@ -324,14 +336,17 @@ func (p *Puller) syncPeer(ctx context.Context, peer swarm.Address, po, d uint8) 
 		binCtx, cancel := context.WithCancel(ctx)
 		syncCtx.binCancelFuncs[uint8(bin)] = cancel
 		if cur > 0 {
+			p.wg.Add(1)
 			go p.histSyncWorker(binCtx, peer, uint8(bin), cur) // start historical
 		}
 		// start live
+		p.wg.Add(1)
 		go p.liveSyncWorker(binCtx, peer, uint8(bin), cur) // start live
 	}
 }
 
 func (p *Puller) histSyncWorker(ctx context.Context, peer swarm.Address, bin uint8, cur uint64) {
+	defer p.wg.Done()
 	if logMore {
 		p.logger.Tracef("histSyncWorker starting, peer %s bin %d cursor %d", peer, bin, cur)
 	}
@@ -387,6 +402,7 @@ func (p *Puller) histSyncWorker(ctx context.Context, peer swarm.Address, bin uin
 }
 
 func (p *Puller) liveSyncWorker(ctx context.Context, peer swarm.Address, bin uint8, cur uint64) {
+	defer p.wg.Done()
 	if logMore {
 		p.logger.Tracef("liveSyncWorker starting, peer %s bin %d cursor %d", peer, bin, cur)
 	}
@@ -431,6 +447,7 @@ func (p *Puller) liveSyncWorker(ctx context.Context, peer swarm.Address, bin uin
 }
 
 func (p *Puller) Close() error {
+	p.logger.Info("puller shutting down")
 	close(p.quit)
 	p.syncPeersMtx.Lock()
 	defer p.syncPeersMtx.Unlock()
@@ -443,6 +460,16 @@ func (p *Puller) Close() error {
 			}
 			peer.Unlock()
 		}
+	}
+	cc := make(chan struct{})
+	go func() {
+		defer close(cc)
+		p.wg.Wait()
+	}()
+	select {
+	case <-cc:
+	case <-time.After(10 * time.Second):
+		p.logger.Warning("puller shutting down with running goroutines")
 	}
 
 	return nil

--- a/pkg/puller/puller_test.go
+++ b/pkg/puller/puller_test.go
@@ -61,7 +61,7 @@ func TestOneSync(t *testing.T) {
 		pullSync: []mockps.Option{mockps.WithCursors(cursors), mockps.WithLiveSyncReplies(liveReplies...)},
 	})
 	defer puller.Close()
-
+	defer pullsync.Close()
 	runtime.Gosched()
 	time.Sleep(10 * time.Millisecond)
 
@@ -193,8 +193,8 @@ func TestSyncFlow_PeerOutsideDepth_Historical(t *testing.T) {
 				},
 				pullSync: []mockps.Option{mockps.WithCursors(tc.cursors), mockps.WithAutoReply(), mockps.WithLiveSyncBlock()},
 			})
-			defer pullsync.Close()
 			defer puller.Close()
+			defer pullsync.Close()
 			runtime.Gosched()
 			time.Sleep(10 * time.Millisecond)
 
@@ -246,8 +246,8 @@ func TestSyncFlow_PeerWithinDepth_Live(t *testing.T) {
 				},
 				pullSync: []mockps.Option{mockps.WithCursors(tc.cursors), mockps.WithLateSyncReply(tc.liveReplies...)},
 			})
-			defer pullsync.Close()
 			defer puller.Close()
+			defer pullsync.Close()
 			runtime.Gosched()
 			time.Sleep(100 * time.Millisecond)
 
@@ -279,8 +279,8 @@ func TestPeerDisconnected(t *testing.T) {
 		pullSync: []mockps.Option{mockps.WithCursors(cursors), mockps.WithLiveSyncBlock()},
 	})
 	t.Cleanup(func() {
-		p.Close()
 		pullsync.Close()
+		p.Close()
 	})
 
 	runtime.Gosched()
@@ -376,8 +376,8 @@ func TestDepthChange(t *testing.T) {
 				},
 				pullSync: []mockps.Option{mockps.WithCursors(tc.cursors), mockps.WithLateSyncReply(tc.syncReplies...)},
 			})
-			defer pullsync.Close()
 			defer puller.Close()
+			defer pullsync.Close()
 
 			runtime.Gosched()
 			time.Sleep(100 * time.Millisecond)

--- a/pkg/pullsync/pullsync.go
+++ b/pkg/pullsync/pullsync.go
@@ -50,6 +50,8 @@ type Syncer struct {
 	streamer p2p.Streamer
 	logger   logging.Logger
 	storage  pullstorage.Storer
+	quit     chan struct{}
+	wg       sync.WaitGroup
 
 	ruidMtx sync.Mutex
 	ruidCtx map[uint32]func()
@@ -71,6 +73,8 @@ func New(o Options) *Syncer {
 		storage:  o.Storage,
 		logger:   o.Logger,
 		ruidCtx:  make(map[uint32]func()),
+		wg:       sync.WaitGroup{},
+		quit:     make(chan struct{}),
 	}
 }
 
@@ -113,15 +117,10 @@ func (s *Syncer) SyncInterval(ctx context.Context, peer swarm.Address, bin uint8
 	}
 
 	ru.Ruid = binary.BigEndian.Uint32(b)
-	defer func() {
-		if err != nil {
-			_ = stream.FullClose()
-			return
-		}
-		_ = stream.Close()
-	}()
 
 	w, r := protobuf.NewWriterAndReader(stream)
+	defer stream.Close()
+
 	if err = w.WriteMsgWithContext(ctx, &ru); err != nil {
 		return 0, 0, fmt.Errorf("write ruid: %w", err)
 	}
@@ -225,6 +224,25 @@ func (s *Syncer) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) err
 		s.ruidMtx.Unlock()
 	}()
 	defer cancel()
+
+	select {
+	case <-s.quit:
+		return nil
+	default:
+	}
+
+	s.wg.Add(1)
+	defer s.wg.Done()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		select {
+		case <-s.quit:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
 
 	var rn pb.GetRange
 	if err := r.ReadMsgWithContext(ctx, &rn); err != nil {
@@ -398,5 +416,17 @@ func (s *Syncer) cancelHandler(ctx context.Context, p p2p.Peer, stream p2p.Strea
 }
 
 func (s *Syncer) Close() error {
+	s.logger.Info("pull syncer shutting down")
+	close(s.quit)
+	cc := make(chan struct{})
+	go func() {
+		defer close(cc)
+		s.wg.Wait()
+	}()
+	select {
+	case <-cc:
+	case <-time.After(10 * time.Second):
+		s.logger.Warning("pull syncer shutting down with running goroutines")
+	}
 	return nil
 }

--- a/pkg/pullsync/pullsync.go
+++ b/pkg/pullsync/pullsync.go
@@ -219,11 +219,15 @@ func (s *Syncer) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) err
 	s.ruidCtx[ru.Ruid] = cancel
 	s.ruidMtx.Unlock()
 	defer func() {
+		select {
+		case <-s.quit:
+			cancel()
+		case <-ctx.Done():
+		}
 		s.ruidMtx.Lock()
 		delete(s.ruidCtx, ru.Ruid)
 		s.ruidMtx.Unlock()
 	}()
-	defer cancel()
 
 	select {
 	case <-s.quit:
@@ -233,16 +237,6 @@ func (s *Syncer) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) err
 
 	s.wg.Add(1)
 	defer s.wg.Done()
-
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	go func() {
-		select {
-		case <-s.quit:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
 
 	var rn pb.GetRange
 	if err := r.ReadMsgWithContext(ctx, &rn); err != nil {

